### PR TITLE
MultipleDatabasesTransitionHelper

### PR DESCRIPTION
**Pull request by Hamms**
Add a helper module for transitioning from the seamless_database_pool gem to ActiveRecord 6.0's new multiple databases support.

## Links

[ActiveRecord 6.0's new Multiple Databases](https://guides.rubyonrails.org/v6.0/active_record_multiple_databases.html)
[seamless_database_pool](https://github.com/bdurand/seamless_database_pool)

## Testing story

I manually tested locally.

First, I created a new 'reader' user on my local database:

```mysql
CREATE USER 'reader'@'localhost' IDENTIFIED WITH mysql_native_password BY '';
GRANT SELECT ON dashboard_development.* TO 'reader'@'localhost';
```

I then added `dashboard_db_reader: "mysql://reader@localhost/dashboard_development"` to my `locals.yml` and enabled mysql's generic query logging by adding to my mysqld.cnf:

```
general_log = 1
log_output = 'table'
```

I then manually overwrote [`ReadReplicaHelper.read_replica?`](https://github.com/code-dot-org/code-dot-org/blob/4191d8657533cd2946c0a3e853e96682ed768d49/dashboard/app/helpers/read_replica_helper.rb#L12) to always return `true`.

Finally, I  visited a script level page (a route that's [explicitly being directed to the reader](https://github.com/code-dot-org/code-dot-org/blob/7335e19d6527b9a9c93fe80f4eac8b73bc474f7c/dashboard/app/controllers/script_levels_controller.rb#L77-L78)) on my local machine, and then consulted the query log table to verify that a subset of queries are made as the read user:

```
mysql> select user_host,count(*) from mysql.general_log GROUP BY user_host;
+-------------------------------+----------+
| user_host                     | count(*) |
+-------------------------------+----------+
| [root] @ localhost []         |        4 |
| root[root] @ localhost []     |     1389 |
| [reader] @ localhost []       |        2 |
| reader[reader] @ localhost [] |       82 |
+-------------------------------+----------+
4 rows in set (0.00 sec)
```

I also verified that requests made to pages without such a configuration do no such splitting.

## Follow-up work

The long-term plan here is:

1. (this PR) Isolate all existing interactions with the SeamlessDatabasePool gem into this helper module, so none of our code outside of this file has specific knowledge of that library
2. Add necessary logic to this helper module for it to dynamically switch from relying on SeamlessDatabasePool to provide the underlying functionality to relying on ActiveRecord.
3. Finish the upgrade from Rails 5.2 to Rails 6
4. Remove this module, replace it with vanilla ActiveRecord implementation

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/helpers/multiple_databases_transition_helper.rb [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+  # Whether we are still on the old version (false, pre-transition) or the new
+  # version (true, post-transition). Hardcoded to the known versions for
+  # resiliency and narrow focus.
+  def self.transitioned?
+    case Rails.version
+    when "5.2.4.4"
+      false
+    when "6.0.4.1"
+      true
```
Comment (cat5inthecradle) [inline dashboard/app/helpers/multiple_databases_transition_helper.rb:67]: love this - I was expecting to just see `transitioned = false`

File: dashboard/app/controllers/application_controller.rb [hunk 2]
```diff
@@ -9,7 +9,7 @@ class ApplicationController < ActionController::Base
   include ApplicationHelper
 
   include SeamlessDatabasePool::ControllerFilter
```
Comment (sureshc) [inline dashboard/app/controllers/application_controller.rb:11]: Why do we retain this usage of Seamless here?

File: dashboard/app/helpers/multiple_databases_transition_helper.rb (hunk 1 above)
Comment (sureshc) [reply dashboard/app/helpers/multiple_databases_transition_helper.rb:67]: Is there a possibility we will patch Rails 5.2 before upgrading to Rails 6 or that we would adopt a slightly different version of Rails 6? Wondering whether we should make the check less specific and distinguish between `5.2.*` vs `6.0.*`

Comment (Hamms) [reply dashboard/app/helpers/multiple_databases_transition_helper.rb:67]: theoretically, yes, but I'm not confident in the ability of this code to support other versions of Rails; also, the idea is that this code should not need to exist for long enough for that to be relevant. If we do end up making unexpected version changes while this code is still around, I want to be alerted to that fact; as I alluded to in the comment, this logic intentionally maintains a narrow focus.

File: dashboard/app/controllers/application_controller.rb (hunk 2 above)
Comment (Hamms) [reply dashboard/app/controllers/application_controller.rb:11]: I attempted to have `MultipleDatabasesTransitionHelper::ControllerFilter` extend from it so we could just include the one, but ControllerFilter is doing some funky stuff internally with its inheritance and initialization logic and I couldn't get it to work. I'm sure it'd be possible, but in the interests of time I decided to leave it as-is, and will just remove this line as part of the other removal work.

**Comment by sureshc** [comment]
@wjordan Check it out! Elijah deployed this tricky change to get us ready to upgrade to Rails 6. Deployed last week smoothly  with no errors!

**Pull request merged**